src/query/storages/fuse/src/operations/mutation/recluster_mutator.rs

```diff
@@ -110,104 +134,210 @@ impl ReclusterMutator {
-                });
-            }
-            assert!(unfinished_parts.is_empty());
-            assert!(!max_points.is_empty());
-
-            let sum_depth: usize = block_depths.iter().sum();
-            // round the float to 4 decimal places.
-            let average_depth =
-                (10000.0 * sum_depth as f64 / block_depths.len() as f64).round() / 10000.0;
-            debug!(
-                "recluster: average_depth: {} in level {}",
-                average_depth, level
-            );
-            if average_depth <= self.depth_threshold {
+            let selected_idx = Self::fetch_max_depth(points_map, self.depth_threshold);
+            if selected_idx.is_empty() {
+                remained_blocks.extend(block_metas.into_iter());
                 continue;
             }
 
-            // find the max point, gather the blocks.
-            let mut selected_idx = HashSet::new();
-            point_overlaps[max_points[0]].iter().for_each(|idx| {
-                selected_idx.insert(*idx);
-            });
+            let blocks_idx: IndexSet<usize> = IndexSet::from_iter(0..len);
+            let diff = blocks_idx.difference(&selected_idx);
+            diff.into_iter()
+                .for_each(|v| remained_blocks.push(block_metas[*v].clone()));
 
+            let mut over_memory = false;
             for idx in selected_idx {
-                let (block_idx, block_meta) = block_metas[idx].clone();
+                let block_meta = block_metas[idx].clone();
+                if over_memory {
+                    remained_blocks.push(block_meta);
+                    continue;
+                }
+
                 let memory_usage = self.total_bytes + block_meta.block_size as usize;
-                if memory_usage > self.memory_threshold {
-                    break;
+                if memory_usage > memory_threshold {
+                    remained_blocks.push(block_meta);
+                    over_memory = true;
+                    continue;
                 }
 
-                self.mutation_logs
-                    .entries
-                    .push(MutationLogEntry::DeletedBlock { index: block_idx });
                 self.total_rows += block_meta.row_count as usize;
                 self.total_bytes = memory_usage;
                 self.selected_blocks.push(block_meta);
             }
 
             self.level = level;
-            return Ok(true);
+            selected = true;
+        }
+
+        if selected {
+            self.remained_blocks = remained_blocks;
+
+            selected_indices.sort_by(|a, b| b.cmp(a));
+            self.removed_segment_indexes = selected_indices;
+
+            let default_cluster_key_id = Some(self.cluster_key_id);
+            selected_statistics.iter().for_each(|v| {
+                merge_statistics_mut(&mut self.removed_segment_summary, v, default_cluster_key_id)
+            });
         }
+        Ok(selected)
+    }
 
-        Ok(false)
+    pub fn select_segments(
```

Comment:
An extreme situation, if a segment's min and max is same, should we skip it? just like block level = -1